### PR TITLE
Allow/ignore upgrades with bodies 

### PR DESCRIPTION
**Pull request by Tratcher**
#17081

Kestrel has had a long behavior of refusing requests (400) if they contain both a Connection: Upgrade header and a request body. The RFCs allow this combination, but you must consume the request body before the upgrade can be completed. See https://github.com/dotnet/aspnetcore/issues/17081#issuecomment-726300552 for details.

Problems:
- Many recommended NGinx configurations always add the Connection: Upgrade header just in case it's needed. This causes Kestrel to refuse any request from NGinx that also has a body.
- It's also possible to use Upgrade to negotiate an optional HTTP/2 upgrade, though ALPN is usually preferred. Clients that do support this will add the headers only to have request with bodies refused by Kestrel.

Proposal: Ignore the `Connection: Upgrade` header if the request has a body, process the requests as normal. This makes Kestrel more compatible without doing the more complex work to fully support upgrade requests that do have bodies.

If we're happy with this fix then I also propose backporting it to 3.1 and 5.0, we've had a lot of feedback about it. And 2.1?

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/KestralBadHttpRequestException.cs [hunk 1]
```diff
@@ -88,9 +88,6 @@ internal static BadHttpRequestException GetException(RequestRejectionReason reas
                 case RequestRejectionReason.InvalidHostHeader:
                     ex = new BadHttpRequestException(CoreStrings.BadRequest_InvalidHostHeader, StatusCodes.Status400BadRequest, reason);
                     break;
-                case RequestRejectionReason.UpgradeRequestCannotHavePayload:
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/src/KestralBadHttpRequestException.cs:91]: Nit: can we fix the file name?

**Review by BrennanConroy — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/KestralBadHttpRequestException.cs (hunk 1 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/KestralBadHttpRequestException.cs:91]: 😆 "Kestral" is how you pronounce it when you want to sound fancy.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/UpgradeTests.cs [hunk 2]
```diff
@@ -164,23 +179,34 @@ await using (var server = new TestServer(context => Task.CompletedTask, new Test
...
-                        "Content-Length: 0",
-                        "",
-                        "");
+                    await connection.Receive("HTTP/1.1 200 OK");
+                }
+
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.SendEmptyGetWithUpgrade();
+                    await connection.Receive("HTTP/1.1 200 OK");
```
Comment (halter73) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/UpgradeTests.cs:190]: Why was this added to the test?

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/UpgradeTests.cs (hunk 2 above)
Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/UpgradeTests.cs:190]: Copied from another upgrade test. At one point I think there was a bug in Kestrel's upgrade logic that broke subsequent requests. These were added as regression tests.

Comment (halter73) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/UpgradeTests.cs:190]: Maybe add a comment.

Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/UpgradeTests.cs:190]: I'll just remove it. It's covered by other tests.
